Sprint 05

bva23 opened the pull request:
Выполнена задача по спринту 5.

bva23 committed:
Выполнены задания до Спринт 5/14: 5 → Тема 2/4: Ответственность → Урок 7/7

bva23 committed:
Выполнено "Задание по теме" Спринт 5/14: 5 → Тема 2/4: Ответственность → Урок 7/7

bva23 committed:
Реализовано свойство isEnabled для кнопок Да/Нет

bva23 committed:
Выполнена задача по спринту 5.

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -84,24 +55,78 @@ final class MovieQuizViewController: UIViewController {
     }
```
В DispatchQueue.main.asyncAfter сильная ссылка, необходимо добавить weak self

protos17 commented on MovieQuiz/Models/QuizQuestion.swift:
```diff
@@ -0,0 +1,14 @@
+
+struct QuizQuestion {
+    let image: String
+    let text: String
```
У тебя повторяется вопрос, рекомендую написать его сразу здесь

protos17 commented on MovieQuiz/Models/QuizStepViewModel.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by Владимир Богомолов on 15.03.2023.
+//
+
+import Foundation
```
При иморте UIKit можно не импортировать Foundation, потому что первый содержит в себе второй

protos17 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,28 @@
+//  Created by Владимир Богомолов on 19.03.2023.
+//
+
+import Foundation
```
Можно убрать Foundation

protos17 commented on MovieQuiz/Presentation/AlertPresenter.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+import UIKit
+
+class AlertPresenter: AlertPresenterProtocol {
```
Давай проставлять ключевое слово `final` классам для ускорения работы компилятора.
Это нужно для оптимизация работы компилятора: ты явно сообщаешь о том, 
что класс не планируется расширять наследованием.

protos17 commented on MovieQuiz/Presentation/AlertPresenterDelegate.swift:
```diff
@@ -0,0 +1,13 @@
+//  Created by Владимир Богомолов on 19.03.2023.
+//
+
+import Foundation
```
Можно удалить Foundation

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -84,24 +55,78 @@ final class MovieQuizViewController: UIViewController {
+        statisticService?.store(correct: correctAnswers, total: questionsAmount)
+
+        guard let statisticService = statisticService else {
+            return ()
```
Лучше написать просто return

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -84,24 +55,78 @@ final class MovieQuizViewController: UIViewController {
+            return ()
+        }
+        guard let bestGame = statisticService.bestGame else {
+            return ()
```
Здесь рекомендую тоже просто return

protos17 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -84,24 +55,78 @@ final class MovieQuizViewController: UIViewController {
+        """
+        Ваш результат: \(correctAnswers)/10
+        Количество сыгранных квизов: \(statisticService.gamesCount)
+        Рекорд: \(bestGame.correct)/10 (\(Date()))
```
Здесь нужно написать Date через DateFormatter

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,34 @@
+
+import Foundation
+
+class QuestionFactory: QuestionFactoryProtocol {
```
Рекомендую указать как final

protos17 commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,34 @@
+        guard let index = (0..<questions.count).randomElement() else {
+            return
+        }
+        let question = questions[safe: index]
```
Рекомендую здесь добавить проверку, что этот вопрос уже не показывался

protos17 commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,95 @@
+            }
+            set {
+                guard let data = try? JSONEncoder().encode(newValue) else {
+                    print("Невозможно сохранить результат")
```
Здесь рекомендую print заменить на assertionFailure

bva23 committed:
Выполнены доработки после ревью:
1. Исправлен формат даты в алерте с окончанием раунда.
2. Сильная ссылка в методе showAnswerResult класса MovieQuizController заменена на слабую.
3. В методе showFinalResults "return ()" заменен на "return".
4. В файле StatisticService.swift при объявлении переменной bestGame вызов сообщения об ошибке заменен с "print" на "assertionFailure".
5. Удалены мок-данные из проекта.
6. Классам проставлен признак final.
7. В файлах, где выполняется импорт UIKit, импорт Foundation теперь отдельно не вызывается.
